src/composables/trade/useJoinExit.ts

```diff
@@ -0,0 +1,318 @@
+  );
+  const { addTransaction } = useTransactions();
+  const { txListener } = useEthers();
+  const { fNum2 } = useNumbers();
+
+  const hasValidationError = computed(
+    () => state.validationErrors.highPriceImpact != null
+  );
+
+  function resetState() {
+    state.validationErrors.highPriceImpact = false;
+
+    state.submissionError = null;
+  }
+
+  function resetInputAmounts(amount: string): void {
+    tokenInAmountInput.value = amount;
+    tokenOutAmountInput.value = amount;
+    priceImpact.value = 0;
+  }
+
+  async function getSwapInfo(): Promise<void> {
+    swapInfoLoading.value = true;
+    swapInfo.value = await balancer.sor.getSwaps(
+      tokenInAddressInput.value,
+      tokenOutAddressInput.value,
+      exactIn.value ? SwapTypes.SwapExactIn : SwapTypes.SwapExactOut,
+      parseFixed(
+        tokenInAmountInput.value || tokenOutAmountInput.value || '0',
+        18
+      ),
+      undefined,
+      true
+    );
+    swapInfoLoading.value = false;
+  }
+
+  async function handleAmountChange(): Promise<void> {
+    // Prevent quering undefined input amounts
+    if (
+      (exactIn.value && !tokenInAmountInput.value) ||
+      (!exactIn.value && !tokenOutAmountInput.value)
+    ) {
+      return;
+    }
+
+    if (pools.value.length === 0) return;
+
+    const amountToExchange = exactIn.value
+      ? tokenInAmountScaled.value
+      : tokenOutAmountScaled.value;
+
+    if (amountToExchange === undefined) {
+      return;
+    }
+
+    if (amountToExchange.isZero()) {
+      tokenInAmountInput.value = '0';
+      tokenOutAmountInput.value = '0';
+      return;
+    }
+
+    await getSwapInfo();
+
+    const tokenInDecimals = getTokenDecimals(tokenInAddressInput.value);
+    const tokenOutDecimals = getTokenDecimals(tokenOutAddressInput.value);
+
+    if (exactIn.value) {
+      tokenOutAmountInput.value = bnum(
+        formatUnits(swapInfo.value?.returnAmount ?? '0', tokenOutDecimals)
+      ).toFixed(6, OldBigNumber.ROUND_DOWN);
+    } else {
+      tokenInAmountInput.value = bnum(
+        formatUnits(swapInfo.value?.returnAmount ?? '0', tokenInDecimals)
+      ).toFixed(6, OldBigNumber.ROUND_DOWN);
+    }
+  }
+
+  async function trade(successCallback?: () => void) {
+    try {
+      confirming.value = true;
+      state.submissionError = null;
+
+      if (!swapInfo.value) {
+        return;
+      }
+
+      const relayerCallData = buildRelayerCalls(
+        swapInfo.value,
+        pools.value,
+        account.value,
+        balancer.contracts.relayerV4?.address ?? '',
+        balancer.networkConfig.addresses.tokens.wrappedNativeAsset,
+        String(slippageBufferRate.value * 1e4),
+        relayerSignature.value || undefined
+      );
+
+      const signer = getSigner();
+      const relayerContract = await balancer.contracts.relayerV4?.connect(
+        signer
+      );
+      const tx = await relayerContract?.multicall(relayerCallData.rawCalls);
```

Comment:
All transactions **must** go through the TxBuilder. We handle a bunch of things there around gas settings etc. Search for uses of `txBuilder.contract.sendTransaction` to see how to construct the call in this scenario.

Reply:
Thanks, updated.